### PR TITLE
Docker image fix

### PR DESCRIPTION
**Pull request by swampie**
This PR allows to run documentation locally by building using mkdocs + insiders gitea hosted plugin.
Follow DEV.md file for details

### CODE REVIEW

**Review by pditommaso — COMMENTED** [review]
worth adding `dockerignore` file to avoid including all content in the container

File: DEV.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# How to build/run documentation locally
+
+## Generate a token on git.seqera.io
+In order to build the docker container to serve the mkdocs site is necessary to generate a token on gitea.
+- Go to [git.seqera.io](https://git.seqera.io) and use your credentials to access
+- In the top right menu click on your avatar and click on `Settings` from the dropdown menu
+- On the top part of the page select `Applications`
+- Generate a new token and copy the value after form submission
+- Put the token into a env variable GITEA_TOKEN and/or store it in .bashrc/.zshrc
```
Comment (llewellyn-sl) [inline DEV.md:9]:
```suggestion
- Put the token into a env variable GITEA_TOKEN and/or store it in .bashrc/.zshrc. Run `source .bashrc` or `source .zshrc` once saved. 
```

File: DEV.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+## Build docs image
+
+
+Run the following command to build the `seqera-docs` image
```
Comment (llewellyn-sl) [inline DEV.md:14]:
```suggestion
Build the `seqera-docs` image:
```

File: DEV.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+`docker build --build-arg GITEA_TOKEN=${GITEA_TOKEN}-t seqera-docs .`
+
+Above command can be replaced by running `make build-docker`
```
Comment (llewellyn-sl) [inline DEV.md:18]:
```suggestion
The above command can be replaced with `make build-docker`. 
```

File: DEV.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+Once command above has been run successfully use the command below to serve the documentation locally
+`docker run --rm -p 8000:8000 -v ${PWD}:/docs seqera-docs:latest serve --dev-addr=0.0.0.0:8000`
+
+
+Above command can be replaced by running `make serve`
+
+
+
+Visit 0.0.0.0:8000 from your favourite browser
```
Comment (llewellyn-sl) [inline DEV.md:30]:
```suggestion
Visit 0.0.0.0:8000 from your favorite browser.
```

**Review by llewellyn-sl — APPROVED** [review]
Made minor suggestions on dev.md. Docker build snd make serve work locally. The rest LGTM!

**Comment by llewellyn-sl** [comment]
I added a .dockerignore, please add to the list of ignores.

**Comment by pditommaso** [comment]
I've made a few changes: 

- improved the `Dockerfile` and `.gitignore` 
- move the dev.md into README.md
- add the the Makefile `docker-image` target that pushes the image to `cr.seqera.io/public/seqera`

It works mostly fine, apart: 

1. `make serve` shows this warning 

```
INFO     -  DeprecationWarning: 'cgi' is deprecated and slated for removal in Python 3.13
              File "/usr/local/lib/python3.11/warnings.py", line 514, in _deprecated
                warn(msg, DeprecationWarning, stacklevel=3)
              File "/usr/local/lib/python3.11/site-packages/encutils/__init__.py", line 55, in
                import CGI
```


2. `make build` crashes 

```
 - installation/system-deployment.md
ERROR    -  Error reading page 'index.md': expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/usr/local/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mkdocs/__main__.py", line 250, in build_command
    build.build(cfg, dirty=not clean)
  File "/usr/local/lib/python3.11/site-packages/mkdocs/commands/build.py", line 308, in build
    _populate_page(file.page, config, files, dirty)
  File "/usr/local/lib/python3.11/site-packages/mkdocs/commands/build.py", line 177, in _populate_page
    page.markdown = config.plugins.run_event(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mkdocs/plugins.py", line 520, in run_event
    result = method(item, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/material/plugins/social/plugin.py", line 125, in on_page_markdown
    page.meta["meta"] = meta + self.__generate_meta(page, config)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/material/plugins/social/plugin.py", line 228, in __generate_meta
    url = "{}.png".format(os.path.join(
                          ^^^^^^^^^^^^^
  File "<frozen posixpath>", line 76, in join
TypeError: expected str, bytes or os.PathLike object, not NoneType
make: *** [build] Error 1
```

**Comment by llewellyn-sl** [comment]
@swampie @pditommaso is further action required here before we can merge this?

**Comment by pditommaso** [comment]
It would be nice to address these errors above. My python skills are too basic to fix it

**Comment by swampie** [comment]
@pditommaso related to serve which is the main goal of this issue (having a running local installation) I am not sure we can fix the warning as there is quite a dependency hell in mkdocs + insiders + python. For the build part consider that this is performed by the github action so not really needed for running the docs locally.


My suggestion is that we move this forward and then also remove insiders as a dependency as the only needed feature is the social card when linking the docs which is already included in newer versions of free mkdocs

**Comment by pditommaso** [comment]
I'm sure @bebosudo can solve this

**Review by bebosudo — CHANGES_REQUESTED** [review]
The problem was the lack of an entry for `site_url`. I added the following to `mkdocs.yml`:
```diff
diff --git a/mkdocs.yml b/mkdocs.yml
index 5d6d5ce..91b69ce 100644
--- a/mkdocs.yml
+++ b/mkdocs.yml
@@ -1,6 +1,7 @@
 site_name: "Tower User Documentation"
 site_author: Seqera Labs
 repo_name: nf-tower-docs
+site_url: https://help.tower.nf
 repo_url: https://github.com/seqeralabs/nf-tower-docs
 edit_uri: https://github.com/seqeralabs/nf-tower-docs/tree/master/docs
```

And now the docs compile without any problem, and without installing the `insiders` nor the libcairo .deb packages.
Could you please address this and the other comments in the review @llewellyn-sl?

**Comment by bebosudo** [comment]
Seems like the social plugin was not part of mkdocs-material version `7.3.6`, which we're hardcoding in the image, for which I'm getting the error:

```
root@9ea5c89f8856:/docs# mkdocs build
ERROR    -  Config value 'plugins': The "social" plugin is not installed
```

I tried removing that hardcoded version, and mkdocs-material was updated to `9.1.15`, which requests the libcairo2 libraries (but not the devel version, which takes python 3.9 with it): I installed libcairo2, and the build process continued but crashed due to [some icons part of fontawesome being renamed](https://fontawesome.com/docs/web/setup/upgrade/whats-changed#icons-renamed-in-version-6), I updated them with the new names:

```diff
@ mkdocs.yml:36 @ theme:
  font: falsediff --git a/mkdocs.yml b/mkdocs.yml
index 5d6d5ce..11aeb1a 100644
--- a/mkdocs.yml
+++ b/mkdocs.yml
@@ -1,6 +1,7 @@
 site_name: "Tower User Documentation"
 site_author: Seqera Labs
 repo_name: nf-tower-docs
+site_url: https://help.tower.nf
 repo_url: https://github.com/seqeralabs/nf-tower-docs
 edit_uri: https://github.com/seqeralabs/nf-tower-docs/tree/master/docs
 
@@ -32,13 +33,13 @@ theme:
   font: false
   icon:
     admonition:
-      note: fontawesome/solid/sticky-note
+      note: fontawesome/solid/note-sticky
       abstract: fontawesome/solid/book
-      info: fontawesome/solid/info-circle
+      info: fontawesome/solid/circle-info
       tip: fontawesome/solid/bullhorn
       success: fontawesome/solid/check
-      question: fontawesome/solid/question-circle
-      warning: fontawesome/solid/exclamation-triangle
+      question: fontawesome/solid/circle-question
+      warning: fontawesome/solid/triangle-exclamation
       failure: fontawesome/solid/bomb
       danger: fontawesome/solid/skull
       bug: fontawesome/solid/robot
```

And now the docs build fine with mkdocs-material 9.

Moreover, what is the `requirements.txt` file used for, since we're manually entering dependencies? Should we remove it, or maybe update it and use it in the build process?

Let me know if you want me to push some commits with the suggestions I made above.

**Comment by swampie** [comment]
@bebosudo Thanks for looking into this: 

my two :coin: 

for me bumping directly to mkdocs >9 was not the original scope of the issue. Insiders is still needed because I think that social plugin is implemented in mkdocs-material >8.
With the latest changes we are now in a situation where local docker execution is different from the publishing process.

I would just implement the change you commented [here](https://github.com/seqeralabs/nf-tower-docs/pull/345#pullrequestreview-1464769339) and then move on with the rest in a subsequent iteration.

Again I am fine with that and I would have done the same but there are too many changes IMHO that could create too much gap between publish and local execution: I will defer decision to @llewellyn-sl and @pditommaso.

**Comment by pditommaso** [comment]
I agree, we should keep this image as much as possible similar to the one used for the publishing, therefore keep material 8 and use `requirements.txt` 

https://github.com/seqeralabs/nf-tower-docs/blob/450621609d7b11787f9aac1e7d7a0a8a22d07765/.github/workflows/staging-html.yml#L45-L47

**Comment by bebosudo** [comment]
> for me bumping directly to mkdocs >9 was not the original scope of the issue

I was suggesting that because most of the errors [Paolo was seeing](https://github.com/seqeralabs/nf-tower-docs/pull/345#issuecomment-1569848174) probably came from using [mkdocs-material 7.3.6 (from October 2021)](https://pypi.org/project/mkdocs-material/7.3.6/) along with the latest [mkdocs 1.4.3 (from May 2023)](https://pypi.org/project/mkdocs/1.4.3/) package.

I tried using the requirements file only, and it kept crashing when using the python:3.11-slim or python:3.11 images: with the former image it was failing due to missing gcc/g++ compilers. Even after adding those and `libffi-dev`, it was still failing when building pyzmq due to a missing header
```
      creating build/temp.linux-x86_64-cpython-311/zmq/backend
      creating build/temp.linux-x86_64-cpython-311/zmq/backend/cython
      gcc -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_SYS_UN_H=1 -Ibundled/zeromq/include -Izmq/utils -I/usr/local/include/python3.11 -c zmq/backend/cython/_device.c -o build/temp.linux-x86_64-cpython-311/zmq/backend/cython/_device.o
      zmq/backend/cython/_device.c:217:12: fatal error: longintrepr.h: No such file or directory
        217 |   #include "longintrepr.h"
            |            ^~~~~~~~~~~~~~~
```

Even though the header is in the image
```
root@2604fe6b9519:/docs# find / -name longintrepr.h
/usr/local/include/python3.11/cpython/longintrepr.h
```

Adding it to CPATH doesn't help much:
```
root@2604fe6b9519:/docs# CPATH=/usr/local/include/python3.11/cpython/ pip install -r requirements.txt 
...
      creating build/temp.linux-x86_64-cpython-311/zmq
      creating build/temp.linux-x86_64-cpython-311/zmq/backend
      creating build/temp.linux-x86_64-cpython-311/zmq/backend/cython
      gcc -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_SYS_UN_H=1 -Ibundled/zeromq/include -Izmq/utils -I/usr/local/include/python3.11 -c zmq/backend/cython/_device.c -o build/temp.linux-x86_64-cpython-311/zmq/backend/cython/_device.o
      zmq/backend/cython/_device.c: In function ‘__Pyx_InitCachedConstants’:
      zmq/backend/cython/_device.c:2793:259: warning: passing argument 14 of ‘PyCode_New’ makes pointer from integer without a cast [-Wint-conversion]
       2793 |   __pyx_codeobj__2 = (PyObject*)__Pyx_PyCode_New(3, 0, 4, 0, CO_OPTIMIZED|CO_NEWLOCALS, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple_, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_zmq_backend_cython__device_pyx, __pyx_n_s_device, 36, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__2)) __PYX_ERR(0, 36, __pyx_L1_error)
...
   In file included from /usr/local/include/python3.11/Python.h:71,
                       from zmq/backend/cython/_device.c:25:
      /usr/local/include/python3.11/cpython/code.h:148:28: note: declared here
        148 | PyAPI_FUNC(PyCodeObject *) PyCode_New(
            |                            ^~~~~~~~~~
      zmq/backend/cython/_device.c: In function ‘__Pyx_AddTraceback’:
      zmq/backend/cython/_device.c:390:62: error: invalid use of incomplete typedef ‘PyFrameObject’ {aka ‘struct _frame’}
        390 |   #define __Pyx_PyFrame_SetLineNumber(frame, lineno)  (frame)->f_lineno = (lineno)
            |                                                              ^~
      zmq/backend/cython/_device.c:5189:5: note: in expansion of macro ‘__Pyx_PyFrame_SetLineNumber’
       5189 |     __Pyx_PyFrame_SetLineNumber(py_frame, py_line);
            |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      ************************************************
      ************************************************
      ************************************************
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyzmq
Failed to build pyzmq
```

It looks to me like that specific version of pyzmq doesn't like something of the new python 3.11. Probably a newer version of that library pyzmq would also work fine on py 3.11.

I tried with `python:3.9`, and it worked flawlessly.
So I think we should stick to that, since we want to keep our GHA and local docker build on the same set of instructions.

@llewellyn-sl here is what I'd use as a Dockerfile, therefore:
```
FROM python:3.9
ARG GITEA_TOKEN

RUN pip install -r requirements \
      && pip install "git+https://${GITEA_TOKEN}@git.seqera.io/seqeralabs/mkdocs-material-insiders@7.3.3-insiders-3.1.3"

WORKDIR /docs
ENTRYPOINT ["mkdocs"]
```

Size could be reduced by using a multi-stage build, first using the complete `python:3.9` image to install all requirements, and then taking the binaries produced and plugging them into a `python:3.9-slim` image.
I can't test it directly since I don't have access to that insiders package.

The best option to keep our GHA and local build in sync would be to use the same container image for both, without installing all dependencies in GHA from scratch, and using a docker/podman action inside the GHA (and it's probably faster too). But we can work on this in another PR.

**Comment by pditommaso** [comment]
Nice. Please update the Dockerfile accordingly

**Comment by pditommaso** [comment]
Ok, there was a glitch but now it works 👍

**Pull request merged**